Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc

```diff
@@ -0,0 +1,177 @@
+
+#include "DataFormats/CTPPSDetId/interface/TotemRPDetId.h"
+#include "DataFormats/CTPPSDetId/interface/TotemTimingDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSPixelDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSDiamondDetId.h"
+
+#include "DetectorDescription/DDCMS/interface/DDCompactView.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/DetGeomDescBuilder.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSDDDNames.h"
+
+#include <regex>
+
+/**
+ * \brief Builds ideal, real and misaligned geometries.
+ *
+ * First, it creates a tree of DetGeomDesc from DDCompView. For real and misaligned geometries,
+ * it applies alignment corrections (RPAlignmentCorrections) found in corresponding ...GeometryRecord.
+ *
+ * Second, it creates CTPPSGeometry from DetGeoDesc tree.
+ **/
+
+using RotationMatrix = ROOT::Math::Rotation3D;
+using Translation = ROOT::Math::DisplacementVector3D<ROOT::Math::Cartesian3D<double>>;
+
+PPSGeometryESProducer::PPSGeometryESProducer(const edm::ParameterSet& iConfig)
+    : verbosity_(iConfig.getUntrackedParameter<unsigned int>("verbosity")),
+      detectorToken_{setWhatProduced(this, &PPSGeometryESProducer::produceIdealGD)
+                         .consumes<cms::DDDetector>(edm::ESInputTag("" /*optional module label */,
+                                                                    iConfig.getParameter<std::string>("detectorTag")))},
+      gdRealTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceRealGD)},
+      gdMisTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedGD)},
+      dgdRealToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceRealTG).consumes<DetGeomDesc>(edm::ESInputTag())},
+      dgdMisToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedTG).consumes<DetGeomDesc>(edm::ESInputTag())} {
+}
+
+void PPSGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<unsigned int>("verbosity", 1);
+  desc.add<std::string>("detectorTag", std::string());
+  descriptions.add("PPSGeometryESProducer", desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+/*
+ * Apply alignments by doing a BFS on idealGD tree.
+ */
+void PPSGeometryESProducer::applyAlignments(const DetGeomDesc& idealGD,
+                                            const CTPPSRPAlignmentCorrectionsData* alignments,
+                                            DetGeomDesc*& newGD) {
+  newGD = new DetGeomDesc(idealGD);
+  std::deque<const DetGeomDesc*> buffer;
+  std::deque<DetGeomDesc*> bufferNew;
+  buffer.emplace_back(&idealGD);
+  bufferNew.emplace_back(newGD);
+
+  while (!buffer.empty()) {
+    const DetGeomDesc* sD = buffer.front();
+    DetGeomDesc* pD = bufferNew.front();
+    buffer.pop_front();
+    bufferNew.pop_front();
```

Comment:
I'm not sure I follow the logic here, does `newGD` get deleted here?

Reply:
`pop_front` does delete the object.

Reply:
Umm, no? `bufferNew` is `std::deque` of raw pointers (so yes, the "pointer object" is deleted, but the pointed-to object is not).

Reply:
Ooops. I am getting confused by this code. What is the lifetime of the object pointed to by `pD`?